### ISSUE TITLE
Creation of a python client

### ISSUE THREAD

**Issue by SystematicError** [localtunnel/localtunnel#394]
Hey! I love this project and regularly use it as an alternative to ngrok to expose my websites built using Flask, a python web framework.

I would love for there to be python client, which I'm willing to make so:

- With your permission, can I build and distribute a python client (proper credit will be given)
- Since I'm one of the least experienced node js programmer out there, could you explain how to tunnel works in order to recreate it using a different language

Thanks in advance!

**Comment by Krish-sysadmin** [comment]
@SystematicError I am sure if you read the source code you will understand how it works. Basically when you type lt --port <yourflaskport> in your console, it starts the localtunnel client. This connects to the express server located in the cloud. This express server returns you the URL where your app should connect. Then it launches the tcp server that listens for connections from your app. The tcp server also opens (I think 10) sockets with the localtunnel client, for sending data. When a connection comes from the app, the tcp server sends the data to the localtunnel client on one of the 10 sockets. The data is then piped to a local HTTP client that issues the request to your server. 

```
 localhost:             [localtunnel client]  --- [HTTP client] --- [your flask server]
                               |                |  
                               |                | 
localtunnel host:      [express server somewhere in ☁️] --- [TCP server]
                                                    |
                                                    |
   internet user:                                 [app you connect to with URL]
   ```

**Comment by SystematicError** [comment]
Thanks for the explanation! I'll attempt to re-write it once I'm free :)

**Comment by mturoci** [comment]
@SystematicError were you successful?

**Comment by SystematicError** [comment]
Well, I don't have any of the code but i was able to use node runtime alongside my python code, other than that, I didn't really get around to implementing a pure python client.

**Comment by mturoci** [comment]
In case anyone is still looking for a python client, I wrote a [blog post](https://medium.com/python-in-plain-english/python-socket-programming-localtunnel-client-11b26ff2ea34) on the matter.

TLDR:

```py
import platform
import urllib.request
import json
import asyncio

# Read from socket1 and write to socket2.
async def pipe(r: asyncio.StreamReader, w: asyncio.StreamWriter) -> None:
    while True:
        # Read 4096 bytes chunk from socket.
        data = await r.read(4096)
        # If EOF is reached, close the pipe.
        if not data:
            break
        w.write(data)
        # Wait until it is appropriate to resume writing to the stream.
        await w.drain()

async def connect(local_host, local_port, remote_host, remote_port):
    # Keep reconnecting if the connection is lost.
    while True:
        local_w = None
        remote_w = None
        try:
            local_r, local_w = await asyncio.open_connection(local_host, local_port)
            remote_r, remote_w = await asyncio.open_connection(remote_host, remote_port)

            # Pipe read-write streams.
            await asyncio.gather(pipe(local_r, remote_w), pipe(remote_r, local_w))
        # Swallow exceptions and reconnect.
        except Exception:
            pass
        finally:
            # Close connection gracefully.
            if local_w:
                local_w.close()
            if remote_w:
                remote_w.close()

async def share(port: int, subdomain: str):
    # Register our app.
    # In real world, you would use
    # a more feature-complete HTTP client (httpx, requests).
    req = urllib.request.Request(
        f'https://localtunnel.me/{subdomain}',
        headers={'Content-Type': 'application/json'}
    )
    response = urllib.request.urlopen(req)
    if response.status != 200:
        print('Could not connect to the remote server.')
        exit(1)

    res = json.loads(response.read().decode())
    
    print(f'Proxying localhost:{port} ==> {res["url"]}')
    print('Press Ctrl+C to stop sharing.')

    connections = []
    # Create concurrent connections.
    for _ in range(res['max_conn_count']):
        connections.append(asyncio.create_task(
            connect('127.0.0.1', port, 'localtunnel.me', res['port']))
        )
    await asyncio.gather(*connections)

def main(port: int, subdomain: str):
    # Just Windows things.
    if 'Windows' in platform.system():
        asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())

    loop = asyncio.new_event_loop()
    asyncio.set_event_loop(loop)

    # Just Windows things.
    if 'Windows' in platform.system():

        async def wakeup():
            while True:
                await asyncio.sleep(1)

        # HACK: Enable Ctrl-C on Windows when opening multiple TCP connections.
        # https://stackoverflow.com/questions/27480967/why-does-the-asyncios-event-loop-suppress-the-keyboardinterrupt-on-windows.
        loop.create_task(wakeup())

    try:
        loop.run_until_complete(share(port, subdomain))
    except KeyboardInterrupt:
        # Shut down gracefully.
        tasks = asyncio.all_tasks(loop)
        for task in tasks:
            task.cancel()
        loop.run_until_complete(asyncio.gather(*tasks, return_exceptions=True))
        loop.close()

if __name__ == '__main__':
    main(10101, 'my-cool-app')
```

**Comment by minch-dev** [comment]
There is still no working python client that well, actually works. The script by @mturoci loads the page once or twice for me then gives 404 and after some time can't even be interrupted, only closing the console works.  https://github.com/jakbin/py-localtunnel "works" in a similar fashion, it just serves exactly one html file and fails to load it's resources, and refreshing it results in 404 error.
